### PR TITLE
REF: move union_categoricals call outside of cython

### PR DESCRIPTION
**Pull request by jbrockmendel**
There's no real perf bump to calling union_categoricals inside cython, better to do it in the python code where we can e.g. get the benefit of mypy. Plus gets us closer to dependency structure goals.

### CODE REVIEW

File: pandas/io/parsers/c_parser_wrapper.py [hunk 1]
```diff
@@ -265,7 +293,76 @@ def _get_index_names(self):
...
+            [
+                f"Columns ({warning_names}) have mixed types."
+                f"Specify dtype option on import or set low_memory=False."
+            ]
+        )
+        warnings.warn(warning_message, DtypeWarning, stacklevel=8)
+    return result
+
+
+def ensure_dtype_objs(dtype):
```
Comment (jreback) [inline pandas/io/parsers/c_parser_wrapper.py:359]: don't we do this more generally elsewhere? e.g. astype should be relocated

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, request

cc @gfyoung

File: pandas/io/parsers/c_parser_wrapper.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/io/parsers/c_parser_wrapper.py:359]: if we did it would probably be in io.common and i dont see it there

**Comment by jbrockmendel** [comment]
gentle ping; lots of annotations to be done around these functions

File: pandas/io/parsers/c_parser_wrapper.py [hunk 2]
```diff
@@ -265,7 +293,76 @@ def _get_index_names(self):
...
+    Concatenate chunks of data read with low_memory=True.
+
+    The tricky part is handling Categoricals, where different chunks
+    may have different inferred categories.
+    """
+    names = list(chunks[0].keys())
+    warning_columns = []
+
+    result = {}
+    for name in names:
```
Comment (jreback) [inline pandas/io/parsers/c_parser_wrapper.py:313]: do we not have a generic pandas concat call to handle this? or is that for a followup

**Review by jreback — CHANGES_REQUESTED** [review]
rebase. also puzzled why the concat logic lives where it does (except i guess historically), is moving this part of the plan?

File: pandas/io/parsers/c_parser_wrapper.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/io/parsers/c_parser_wrapper.py:342]: > do we not have a generic pandas concat call to handle this? or is that for a followup

im hopeful we can use something more generic, but havent tried it out yet (comment on L336)

**Comment by jbrockmendel** [comment]
id like to get this in before the upcoming PR doing annotations for io.parsers, which has recently become a blocker for #41059

File: pandas/_libs/parsers.pyx [hunk 3]
```diff
@@ -774,17 +767,18 @@ cdef class TextReader:
...
 
-    #  -> dict[int, "ArrayLike"]
-    cdef _read_low_memory(self, rows):
+    def read_low_memory(self, rows: int | None)-> list[dict[int, "ArrayLike"]]:
+        """
+        rows=None --> read all rows
+        """
+        # Conserve intermediate space
+        # Caller is responsible for concatenating chunks,
+        #  see c_parser_wrapper._concatenatve_chunks
```
Comment (simonjayhawkins) [inline pandas/_libs/parsers.pyx:781]: typo

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, can you catch @simonjayhawkins comment in next pass.

**Pull request merged**